### ISSUE TITLE
Missing postalcode for street result

### ISSUE THREAD

**Issue by otbutz** [pelias/pelias#789]
Hey team! 

I was using your awesome geocoding engine when I noticed something interesting.
Let me tell you more about it.

---
#### Here's what I did :innocent:
Geocode the following street: Theodor-Storm-Strasse 95032 Hof DE
https://pelias.github.io/compare/#/v1/search%3Ftext=Theodor-Storm-Strasse%2095032%20Hof%20DE

---
#### Here's what I got :scream_cat:

Exact match without the postalcode:

```
{
  "geocoding": {
    "version": "0.2",
    "attribution": "https://geocode.earth/guidelines",
    "query": {
      "text": "Theodor-Storm-Strasse 95032 Hof DE",
      "size": 10,
      "private": false,
      "lang": {
        "name": "German",
        "iso6391": "de",
        "iso6393": "deu",
        "defaulted": false
      },
      "querySize": 20,
      "parser": "libpostal",
      "parsed_text": {
        "street": "theodor-storm-strasse",
        "postalcode": "95032",
        "city": "hof",
        "country": "DEU"
      }
    },
    "engine": {
      "name": "Pelias",
      "author": "Mapzen",
      "version": "1.0"
    },
    "timestamp": 1553788978270
  },
  "type": "FeatureCollection",
  "features": [
    {
      "type": "Feature",
      "geometry": {
        "type": "Point",
        "coordinates": [
          11.900806,
          50.297433
        ]
      },
      "properties": {
        "id": "polyline:4403738",
        "gid": "openstreetmap:street:polyline:4403738",
        "layer": "street",
        "source": "openstreetmap",
        "source_id": "polyline:4403738",
        "name": "Theodor-Storm-Straße",
        "street": "Theodor-Storm-Straße",
        "confidence": 1,
        "match_type": "exact",
        "accuracy": "centroid",
        "country": "Deutschland",
        "country_gid": "whosonfirst:country:85633111",
        "country_a": "DEU",
        "region": "Bayern",
        "region_gid": "whosonfirst:region:85682571",
        "region_a": "BY",
        "macrocounty": "Oberfranken",
        "macrocounty_gid": "whosonfirst:macrocounty:404227583",
        "county": "Hof",
        "county_gid": "whosonfirst:county:102063611",
        "county_a": "HS",
        "locality": "Hof",
        "locality_gid": "whosonfirst:locality:101841157",
        "continent": "Europa",
        "continent_gid": "whosonfirst:continent:102191581",
        "label": "Theodor-Storm-Straße, Hof, Deutschland"
      },
      "bbox": [
        11.899324,
        50.29441,
        11.901414,
        50.300262
      ]
    }
  ],
  "bbox": [
    11.899324,
    50.29441,
    11.901414,
    50.300262
  ]
}
```
---
#### Here's what I was expecting :sparkles:
The postalcode is part of the WOF dataset(last modified: 2016-09-22T18:40:15.000Z) and the street is part of its shape:
https://spelunker.whosonfirst.org/id/421245499/

![grafik](https://user-images.githubusercontent.com/22315436/55173370-e36e6180-517b-11e9-8a05-fc076c04d9f8.png)

---
#### Here's what I think could be improved :trophy:
The postalcode should be part of the returned result.

**Comment by NickStallman** [comment]
This would be similar to my issue #777

The short answer is postcodes are hard. :)

The longer answer is it would be ideal to include the postcode as part of the street, if the street is entirely enclosed by the postcode.
This would be done as part of PIP in Pelias, unlike my issue in #777 which would likely involve Who's on First.

**Comment by otbutz** [comment]
I wholeheartedly agree that these kind of things can get messy very quickly. But the case at hand seems rather trivial. PIP should be able to solve this.

**Comment by missinglink** [comment]
I don't believe these polygons are accurate, from some quick research I can see that 95032 should include `Eppenreuth` which isn't included in this shape.

Other polygon shapes can overlap each other or even cover addresses which are actually a different postcode, they are of poor quality globally and not generally from authoritative sources.

Because they are of variable quality internationally we've decided not to includee them in the point-in-polygon system at this stage.

The tradeoff there is that we *could* assign postcodes to records which don't have it, but a large number of those would be incorrect.
So if we did this we'd instead receieve bug reports about *incorrect* postcode instead of *missing* postcodes and it could hurt the project reputation as we would be providing a mix of accurate and inaccurate data.

**Comment by missinglink** [comment]
I would seriosuly consider adding these shapes to the point-in-polygon service if a subset of them (maybe a whole country?) can be shown to be accurate.

We could then enable them for this smaller subset and work forward to supporting more data as more reliability is confirmed.

**Comment by otbutz** [comment]
Incorrect data should be fixed at its source IMHO. But overlapping postalcode shapes are indeed a problem.

Suggestion:
* use the postalcode of the address if it's already present
* use the postalcode obtained via PIP if it's unambiguous

**Comment by missinglink** [comment]
Yes, I totally agree, we currently already do 1. for addresses (it's included when present in the source data)
A `street` is different from an `address` in that it can easily span multiple postal codes and streets don't commonly have associated postcodes.

A clear example is the [Øresund_Bridge](https://en.wikipedia.org/wiki/%C3%98resund_Bridge) which spans two countries, although it's actually far more common than you might think.

The US Census T.I.G.E.R. dataset encoded block ranges with four zip codes, one for each of the left/right start and left/right end of the street.

I would be interested to hear of any geographies where streets have a postcode, I've not seen this myself but wouldn't be surprised if it were true.

For your second point, my experience of the WOF postcode boundaries is that they are of a low quality internationally, so even if they are unambiguous within the dataset I don't believe they are reliable enough to use due to the reasons I mentioned above.

I think a good next step would be to investigate the postalcode polygons for one single country and evaluate their accuracy.

**Comment by missinglink** [comment]
Another option worth considering is doing a reverse search on the lat/lon of the street, this will return addresses near the street, if the `address` records have a postcode then you can use that

Eg.
> /v1/reverse?layers=address&point.lon=11.900806&point.lat=50.297433

Unfortunately in Hof the address data does not come with postcode information :(

**Comment by otbutz** [comment]
> A `street` is different from an `address` in that it can easily span multiple postal codes and streets don't commonly have associated postcodes.

Correct. It *can* span multiple postal codes. But if the whole street is within a single postal code shape i'd expect the service to return it.

eg Bad Wöhrishofen(DE) has a single postal code: 86825
https://nominatim.openstreetmap.org/search.php?q=86825&polygon_geojson=1&viewbox=

A query for a street which doesn't cross the border of the shape should return results with this postal code.

> For your second point, my experience of the WOF postcode boundaries is that they are of a low quality internationally, so even if they are unambiguous within the dataset I don't believe they are reliable enough to use due to the reasons I mentioned above.

I agree that there might be issues with the WOF data but it should be possible to configure it as a fallback.

**Comment by NickStallman** [comment]
Australia should be a good place to test this functionality, like my other ticket postcodes are (mostly) very well defined and shouldn't have overlaps, and if the source data is inaccurate it's clear cut what the problem is.

Can PIP currently do this? I thought it was only Point in Polygon not Polygon in Polygon?

Since this won't alter the postcode of addresses (which would be bad) this should be relatively safe even if it isn't 100% exact. It's only adding new information that wasn't returned previously.

**Comment by otbutz** [comment]
> Can PIP currently do this? I thought it was only Point in Polygon not Polygon in Polygon?

I *think* it's not supported ATM but it would be a good addition nevertheless. It should in any case be possible to check where the start and end points of the street are.

**Comment by missinglink** [comment]
Hi @NickStallman, check out this for nasty https://mobile.twitter.com/_drei/status/1113293065072939011

**Comment by missinglink** [comment]
I'm giving this some serious thought and I will discuss it with the WhosOnFirst team next time we meet.

My first-hand experience here in Berlin is that German postcodes appear to be conterminous with administrative boundaries, and I think there is supporting evidence that this is true for all of Germany.

So I think Germany would be a good country to start with, although I'd caution that any solution we adopt for Germany will unlikely apply elsewhere as it seems very rare for postcodes to be conterminous in the same way for other countries.

**Comment by missinglink** [comment]
The fundamental question in my mind is how these algorithmically associated postcodes would be presented to the user.

If we were to introduce a new property called something like 'postcode.confidence' and set this lower for 'guessed' postcodes and higher for 'real' postcodes then I would feel much happier adopting solutions which would return false positives.

**Comment by NickStallman** [comment]
@missinglink hahaha that's a great graphic. When the postcode is getting close to the size of some countries it becomes more about which post office is closest even if it's crossing state boundaries. ;)

I think it's important because end users often expect the postcode to be there.
If all edge cases need to be eliminated first then a project like pelias won't go anywhere fast. :)

A flag to show calculated vs actual data is a fantastic compromise. If a use case requires only authoritative data then it allows it to be ignored when it's calculated whilst still allowing it to be useful for use cases where 'close enough is good enough'.

**Comment by jeremy-rutman** [comment]
It looks like the [point-in-polygon api ](http://pelias.github.io/compare/#/v1/reverse%3Fpoint.lon=138.16&point.lat=-19.61) doesnt look at [wof postalcode regions](https://spelunker.whosonfirst.org/id/504781293/)   which iiuc is controlled by the pelias/pip-service `node_modules/pelias-whosonfirst/utils/download_data.js --admin-only`  ; possibly by changing that admin-only to also include postal codes then the postcode info would be returned

**Comment by NickStallman** [comment]
@jeremy-rutman yeah PIP doesn't even try to load the postalcode databases, but the imports.whosonfirst.importPostalcodes = true config option will already download them.
That flag downloads them and adds them to the index for searching, but they aren't used for any other purpose so we'd need another flag to cause PIP to load them and then code to run street points through it.

**Comment by msschl** [comment]
Are there any news on this?

**Comment by divE4pip** [comment]
+1 on this feature
